koocook_core/support/query/ruleset.py

```diff
@@ -1,21 +1,68 @@
+import operator
+from functools import reduce
+from typing import Tuple, Type
+from django.db.models import Q, QuerySet
+from django.http import QueryDict
+from django.core.exceptions import FieldError
+
+
 class Rule:
     name: str = ""
     verbose_name: str = ""
     key: str = ""
-    pass
 
-
-class Ruleset:
     def __init__(self):
-        self.rules = []
+        self.rule_body = None
+
+    def process(self, request_fields: QueryDict, queryset: QuerySet) -> QuerySet:
+        return queryset
 
-    def add_rule(self, rule: Rule):
-        self.rules.append(rule)
+    def validate(self, request_fields: QueryDict, queryset: QuerySet) -> QuerySet:
+        if self.key not in request_fields:
+            return queryset
+        elif request_fields.get(self.key) and request_fields.get(self.key) != '':
+            self.rule_body = request_fields.get(self.key)
+            return self.process(request_fields, queryset)
+        else:
+            return queryset
+
+
+class QueryRuleset:
+    def __init__(self, *args: Type[Rule]):
+        self.rules: Tuple[Type[Rule]] = args
```

Comment:
You sure it's `Tuple[Type[Rule]]` for tuple of length one and not `Tuple[Type[Rule], ...]` of variable length?

Reply:
Does that affect the way type checking behave?

Reply:
Yes it does